### DISCUSSION TITLE
Losing asgiref.Local value with async graphql resolvers

### DISCUSSION THREAD

**Discussion by ZipBrandon** [lorinkoz/django-pgschemas#124] [Q&A]
I am using https://strawberry.rocks/ as my GraphQL library.  Everything works as expected with django-pgschemas when using it synchronously.  `get_current_schema()` will return the schema I activated in my `TenantProtocolRouter` subclass.  When I convert a query resolver to async (by introducing the async keyword to the function definition) then I receive a `<django_pgschemas.schema.SchemaDescriptor object at 0xffff7114ecb0>` when I call `get_current_schema()` and the tenant that I activated doesn't appear to be in the scope of the executing query resolver.  I don't have a great understanding of `asgiref.Local` so I don't know if this would be expected.  I migrated from `django-tenants` to circumvent the storing of the tenant on the `connection` and was hoping `django-pgschemas`'s solution would help.  Is there something that I could be conceptually missing?

**Comment by lorinkoz** [comment]
Hola @ZipBrandon. Can you give me more information so I can try to reproduce?

**Reply by ZipBrandon** [reply]
Hi @lorinkoz, thanks for the response.  I want to acknowledge that I saw this and I'll see if I can create a minimal GitHub repo to try to reproduce what I'm seeing on my side.

**Reply by ZipBrandon** [reply]
@lorinkoz I have created a minimal setup that exemplifies this behavior at https://github.com/ZipBrandon/pgschemas_async.  The README.md has instructions to easily get this up and running. Thank you!

**Reply by lorinkoz** [reply]
On a first look: the call method of the protocol router is not the intended place to activate the tenant, but let me get back to you with a proper explanation.

EDIT: And thank you for taking the time to create a reproducible example.

**Reply by lorinkoz** [reply]
Sorry for the delay. To confirm, the protocol router is not the place to activate the tenant, but the consumer itself. I typically do in the connect of the consumer. TBH, I don't know the full explanation, but AFAIK the consumers operate on a different context and therefore a tenant activated in the router is not shared to the consumers.

Could you please overwrite the graphql consumer, activate the tenant there from the scope, and check if that works?

**Reply by ZipBrandon** [reply]
It's still losing the context of the schema in my resolver.  I have been tracing the call stack to identify where but I have not been able to locate the transition point, unfortunately.  It appears it may be in `graphql-core` but I have not been able to conclusively determine.

**Reply by lorinkoz** [reply]
Alright, my bad. This isn't trivial. Let me take a deeper look.

**Reply by lorinkoz** [reply]
I've been unable to trace the problem. Now, side question: because the tenant successfully arrives to your queries via `info.context`, would it be acceptable as a _temporary_ workaround to activate the tenant right at the query?

**Reply by ZipBrandon** [reply]
@lorinkoz Exactly. You are correct and the ops that have to be `async` are being handled this way.

**Reply by lorinkoz** [reply]
My key takeaway: I need to add proper tests for `django-channels` both sync and async mode, so I can be sure whether this is a `pgschemas` thing or a `strawberry` thing. Let's see if I can allocate some time for that. In the meantime, if you or anyone from your team wants to take a shot at it, that would be highly appreciated.

**Reply by lorinkoz** [reply]
On a separate topic @ZipBrandon, we are using `pgschemas` in a single project that doesn't use the full set of features (namely async), therefore I don't have sufficient confidence that this package is production ready yet. Please use it with a grain of salt, as I know `django-tenants` is battle tested and way more mature than this one.
